File: code/typescript-advanced/compare_fail/L01InferFromName.java
// compare_fail/L01InferFromName.java
import java.util.function.Consumer;

public class L01InferFromName {
    record NavigateToPlanet(String target) {}

    static <T> void on(String name, Consumer<T> handler) {}

    public static void main(String[] args) {
        on("NavigateToPlanet", data -> System.out.println(data.target()));
    }
}

File: code/typescript-advanced/compare_fail/l01_infer_from_name.cs
// compare_fail/l01_infer_from_name.cs
// With a string name, nothing tells the compiler which payload the handler receives
var hub = new Hub();
hub.On("NavigateToPlanet", data => Console.WriteLine(data.Target));

class Hub
{
    public void On<T>(string name, Action<T> handler) { }
}
